fix: register validation error

Claire-bk opened the pull request:
remove not match field from validator

Claire-bk committed:
fix: register validation error

remove not match field from validator

Claire-bk committed:
fix: package.json

modified package.json and update package-lock.json

Claire-bk committed:
fix: double quotes

change single quotes to double quotes

Jet004 commented on __test__/integrationTests.test.ts:
```diff
@@ -252,3 +281,17 @@ describe("Integration tests for AUTH routes:", () => {
     });
 });
+
+interface UserRegister {
```
An identical interface already exists at ./models/userModel.ts called RegistrationData. You can import it and use it directly.

Claire-bk replied:
I've imported it from userModel.

Jet004 commented on middleware/userRegisterValidator.ts:
```diff
@@ -0,0 +1,12 @@
+    email: yup.string().email().trim().required(),
+    password: yup.string().trim().min(4).max(10)
+        // Minimum eight and maximum 10 characters, at least one uppercase letter, one lowercase letter, one number and one special characte
+        // .matches(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{4,10}$/)
```
This file exists already in ./validation/userRegisterValidator.ts. Can you please delete it from ./middleware?

Claire-bk replied:
The validation folder is for all validation files. login, post and other validation files will go there. I will delete the duplicate file from middleware folder.

Jet004 commented on middleware/validator.ts:
```diff
@@ -0,0 +1,13 @@
+import { NextFunction, Request, Response } from "express";
+import { AnyObjectSchema } from "yup";
+
+export const validateMiddleware = (schema: AnyObjectSchema) => {
```
I suggest changing the name of the function *validateMiddleware* to just *validate* as it will be more readable in routes. eg:
```
router.route("/api/auth/register")
    .post(
        validate(registrationSchema),
        register(...)
    )
```

Jet004 commented on routes/authRoutes.ts:
```diff
@@ -11,6 +13,7 @@ router
     .route("/register")
     .post(
         protect({ loggedIn: false }),
+	    validateMiddleware(registerSchema),
```
Make sure all of the functions are in line. validateMiddlware is indented further than the protect functions

Jet004 commented on validation/userRegisterValidator.ts:
```diff
@@ -0,0 +1,12 @@
+    email: yup.string().email().trim().required(),
+    password: yup.string().trim().min(4).max(10)
+        // Minimum eight and maximum 10 characters, at least one uppercase letter, one lowercase letter, one number and one special characte
+        // .matches(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{4,10}$/)
```
As the .matches statement is commented out it is inactive. Is there a reason for this? When this is uncommented it will require users to create strong passwords. Also, we will need more tests to check that this works as intended.

Claire-bk replied:
That was an example. We don't have any rules about password. I thought after discuss about the password rule, I would change it but if you are happy with the rule, I will uncomment it.

Jet004 replied:
I posted it on Discord and it seems most would be happy with the pattern below to begin with. By adding this pattern you no longer need to add .min(4).max(10) as the pattern will also check length.

// Minimum six and maximum 15 characters, at least one uppercase letter, one lowercase letter, one number and one special character
// .matches(/^(?=.[a-z])(?=.[A-Z])(?=.\d)(?=.[-@$!%?&])[A-Za-z\d-@$!%?&]{6,15}$/)

Jet004 commented on validation/userRegisterValidator.ts:
```diff
@@ -0,0 +1,12 @@
+import * as yup from "yup"; 
```
I recommend that this file be renamed to *registrationValidator* to simplify it.

Jet004 reviewed (changes requested):
This is good. Something else we may need to consider is that often we will need to validate more than just req.body. Sometimes we will need to validate req.params or req.headers. We may need extra validation middleware to cover these use cases

Claire-bk committed:
fix:merge conflict

 fix package-lock.json

Claire-bk committed:
fix: pull request

remove userRegister interface, change validateMiddleware to validate, change yup to registrationValidator

Claire-bk committed:
fix: faker

add faker-js

Claire-bk committed:
fix: push package-lock.json

Claire-bk committed:
fix : error

 change single quotes to double

Claire-bk committed:
fix: userRegisterValidator

delete userRegisterValidator.ts from middleware

Jet004 commented on __test__/integrationTests.test.ts:
```diff
@@ -120,6 +156,35 @@ describe("Integration tests for AUTH routes:", () => {
+            const response = await axios.post("api/auth/register", user);
+            expect(response.status).toBe(400);
+        });
+    });
```
This is a good start. We will need to add more tests once the password pattern is in place though to make sure it is working as intended.

Claire-bk replied:
Thanks, Jacob.

Jet004 commented on models/migrations/20220727232515_init/migration.sql:
```diff
@@ -0,0 +1,203 @@
+-- CreateEnum
```
This file should be deleted as per the discussion on Discord: https://discord.com/channels/994239034362691765/995931261422010448/1006390907899228261

Claire-bk replied:
I deleted the file. Thanks.

Jet004 commented on validation/userRegisterValidator.ts:
```diff
@@ -0,0 +1,14 @@
+import * as registrationValidator from "yup";
```
I should be more clear with my comments. My previous comment suggested that this file be renamed to registrationValidator. What I meant was that the file --> ./validation/userRegisterValidator.ts be renamed to ./validation/registrationValidatory.ts
Yup was ok as it way. Sorry for the confusion

Claire-bk replied:
I will change userRegisterValidator.ts to userValidator.ts because it will cover user registration and user login validation. I will push it soon.

Jet004 replied:
Can it cover the use case for both register and login? Register requires 3 fields: userName, email and password. Login only requires 2 fields: email and password.

Claire-bk replied:
I've done it but I didn't push it. I will push it now.

Jet004 commented on validation/userRegisterValidator.ts:
```diff
@@ -0,0 +1,14 @@
+        .min(4)
+        .max(10)
+        // Minimum eight and maximum 10 characters, at least one uppercase letter, one lowercase letter, one number and one special characte
+        // .matches(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,10}$/)
```
I posted it on Discord and it seems most would be happy with the pattern below to begin with. By adding this pattern you no longer need to add .min(4).max(10) as the pattern will also check length.

// Minimum six and maximum 15 characters, at least one uppercase letter, one lowercase letter, one number and one special character
// .matches(/^(?=.[a-z])(?=.[A-Z])(?=.\d)(?=.[-@$!%?&])[A-Za-z\d-@$!%?&]{6,15}$/)

Claire-bk replied:
I changed the pattern and "password" to "Abc-1234" in integrationTests.test.tc as well. 
You can use createRandomUserForRegister() to create random user info.

Claire-bk committed:
fix: delete migration file

 delete _init/migration.sql

Claire-bk committed:
fix: password pattern

Change password pattern and invalid password

Claire-bk committed:
feat: login validation

Add login validation

Claire-bk committed:
fix: password pattern

Change strong password pattern

Claire-bk committed:
feat: post validator

Add post validator

B08waffles reviewed (approved):
Well done. "faker-js" is cool, I'm keen to see this merged!

Jet004 commented on middleware/validator.ts:
```diff
@@ -0,0 +1,25 @@
+    };
+};
+
+export const sessionValidate = (schema: AnyObjectSchema) => {
```
There shouldn't really be any need to validate session data as it is always stored on the backend. Session data is created after user input is validated in the registration route. After that it is saved in the database. Session data can't be modified by users directly, so it should always be in a safe state. We do need to implement CSRF tokens though - we can get to that later.

Jet004 commented on validation/userValidator.ts:
```diff
@@ -0,0 +1,20 @@
+import * as validator from "yup";
+
+const validateLogin = {
+    email: validator.string().email().trim().required(),
```
Rather than validateLogin, I suggest defining email, password etc. separately. This way they can be used in other validation files too. You can even move them to a validationDefinition file and import them where they are needed. For example, in validationDefinition.ts:
```
export const emailValidator = yup.email('error message').min().max().trim().required()
```
Then in userValidator.ts:
```
import { emailValidator } from "./valitationDefinition"

export const registerSchema = yup.object({
    email: emailValidator,
    ... other validators
})
```

This makes all of the fields reusable.

Jet004 commented on __test__/integrationTests.test.ts:
```diff
@@ -120,6 +156,35 @@ describe("Integration tests for AUTH routes:", () => {
+            }
+        );
+
+        test("return 400 when password is too short", async () => {
```
This is good, but there will need to be more tests to make sure the regex pattern works correctly. Eg. password too long, no special character, no capital letters, no lowercase letters, no numbers etc...
email tests for a number of possibly invalid emails
username tests length and data type, test with numbers, short, long etc.

Jet004 commented on validation/userValidator.ts:
```diff
@@ -0,0 +1,20 @@
+
+export const registerSchema = validator.object({
+    ...validateLogin,
+    userName: validator.string().trim().required(),
```
I suggest that min and max be added to the userName validator. also we should define what values can be used. numbers? special characters?

Claire-bk replied:
I will leave a comment on discord.

Claire-bk committed:
fix: session validate

remove ression validate

Claire-bk committed:
fix: user validation

create validation definition file and apply changes on userValidator

Claire-bk committed:
fix: password validation test

add password validation test case

Claire-bk committed:
fix: usernameValidattor

add regex pattern for username

Claire-bk committed:
fix: username validation test

add username validatin test code

Claire-bk committed:
fix: email validation test

add email validation test case

Claire-bk committed:
fix:loger username for test

use loger username to test too long username

Jet004 commented on validation/postValidation.ts:
```diff
@@ -0,0 +1,16 @@
+import * as yup from "yup";
+
+export const validateUserId = {
+    userId: yup.string().trim().required(),
```
Great work. We don't need to validate the userId for posts as we get it directly from session data. For future reference Yup has a uuid checker which will make sure it is actually a uuid rather than just a string: https://github.com/jquense/yup#stringuuidmessage-string--function-schema

Claire-bk replied:
Hi Jet,
I want to push this pull request only for user validation. I will pull request for post validation next time. Can I merge it?

Claire-bk replied:
I changed the code to check if it is uuid or not. I will create another pull request for post validation after fix validatePost.

Jet004 replied:
Ok, that's fine. Let's merge once the tests all pass. I suggest that you create a new branch to keep the changes separate in future, though 😉

Jet004 reviewed (changes requested):
Great work. Only one small point and I think it's ready to be merged!

Claire-bk committed:
fix: uuid validation

change validate user id code to check if it is uudi or not

Claire-bk committed:
fix: userIdValidator

move userIdSchema to validationDefinition.ts

Claire-bk committed:
Merge branch 'development' into claire-validation

Claire-bk committed:
fix: user update password

change invalid password to valid one for user update

Claire-bk commented:
Merge User validation
